src/IROperator.cpp

```diff
@@ -566,6 +569,42 @@ Expr lossless_cast(Type t, Expr e) {
         return Shuffle::make(vecs, shuf->indices);
     }
 
+    if (const Mod *mod = e.as<Mod>()) {
+        Expr val;
+        if (e.type().is_uint()) {
+            val = simplify(mod->b + make_const(e.type(), -1));
+        } else if (e.type().is_int()) {
+            // 0 <= a%b < |b|
+            Type wide_ty = e.type().with_bits(e.type().bits() * 2);
```

Comment:
In the cases where you return cast(t, e), could you just do this:
```
Expr x = cast(t, e);
if (can_prove(cast(e.type(), x) == e) {
  return x;
}
```
?

Reply:
Or maybe
```
if (e.type().can_represent(t.min()) && 
    e.type().can_represent(t.max()) && 
    can_prove(e >= cast(e.type(), t.min()) && 
    e <= cast(e.type(), t.max())) 
```

Basically I'm wondering if we can do something that appeals to verified code to reason about the bounds of mods and ramps, rather than having to worry about this method being correct. It's hard to get bounds right.